pkg/endpoint/policy.go

```diff
@@ -498,14 +512,91 @@ func (e *Endpoint) Regenerate(regenMetadata *regeneration.ExternalRegenerationMe
 				e.getLogger().Debug("regeneration was cancelled")
 			}
 		}
-
-		done <- buildSuccess
-		close(done)
 	}()
 
 	return done
 }
 
+var reasonRegenRetry = "retrying regeneration"
+
+// startAsyncRegenerationFailureHandler launches a goroutine which waits for a build of
+// the Endpoint to fail. If the Endpoint is deleted, the goroutine is terminated.
+// If a build fails, a controller is launched which tries to regenerate the
+// Endpoint until it suceeds. Once the controller succeeds, it will not be
+// ran again unless another build failure occurs.
+func (e *Endpoint) startAsyncRegenerationFailureHandler() {
+	var boff backoff.Exponential
+	var numConsecutiveFailures int
+	go func() {
+		for {
+			select {
+			case <-e.regenFailedChan:
+				e.getLogger().Debug("received signal that regeneration failed")
+				select {
+				case <-e.aliveCtx.Done():
+					e.getLogger().Debug("exiting retrying regeneration goroutine due to endpoint being deleted")
+					return
+				default:
+				}
+			case <-e.aliveCtx.Done():
+				e.getLogger().Debug("exiting retrying regeneration goroutine due to endpoint being deleted")
+				return
+			}
+
+			if numConsecutiveFailures == 0 {
+				// Reset backoff if we've had success between the last time
+				// failure occurred.
+				boff = backoff.Exponential{Min: time.Duration(100) * time.Millisecond}
+			}
+			// Maintain backoff outside of the controller - we need to meter
+			// controller updates here. Calling `UpdateController` will reset
+			// the number of retries if regeneration keeps on failing, so we
+			// limit it ourselves. Starting regeneration itself from a
+			// controller is difficult because a given regeneration takes in
+			// context / is expected to return values to a given caller of
+			// Regenerate(). This could definitely be cleaned up later with some
+			// more fine-tuned combing through of the Endpoint regeneration
+			// code.
+			if err := boff.Wait(e.aliveCtx); err != nil {
+				// Error is only returned if the context is done; in this case
+				// that means the endpoint was deleted, so we can exit
+				// gracefully.
+				return
+			}
+
+			e.controllers.UpdateController(fmt.Sprintf("endpoint-%s-regeneration-recovery", e.StringID()), controller.ControllerParams{
```

Comment:
If an endpoint has not been successfully regenerated in line 587, won't we end up updating this controller again?

Reply:
Yes, we will. Is there a concern you see in how it is structured this way?

Reply:
We could simply leverage the whole logic for the back off already built-in in the controllers logic. We would create the controller on line 532/533 which would still block by the channel on line 540.

Reply:
This is much cleaner. I'm not sure why I didn't do this myself.

Reply:
PTAL